File: src/backendng/src/main/kotlin/com/secman/service/ApplicationRegisterService.kt
package com.secman.service

import com.secman.domain.ApplicationRegister
import com.secman.dto.ApplicationRegisterDetail
import com.secman.dto.ApplicationRegisterRequest
import com.secman.dto.ApplicationRegisterSummary
import com.secman.repository.ApplicationRegisterRepository
import com.secman.repository.AssetRepository
import jakarta.inject.Singleton
import jakarta.transaction.Transactional

@Singleton
open class ApplicationRegisterService(
    private val repository: ApplicationRegisterRepository,
    private val assetRepository: AssetRepository
) {

    @Transactional
    open fun list(search: String?): List<ApplicationRegisterSummary> {
        return repository.search(search?.trim()).map { ApplicationRegisterSummary.from(it) }
    }

    @Transactional
    open fun get(id: Long): ApplicationRegisterDetail {
        return ApplicationRegisterDetail.from(getWithAssets(id))
    }

    @Transactional
    open fun create(request: ApplicationRegisterRequest, username: String): ApplicationRegisterDetail {
        val cleaned = clean(request)
        val carId = allocateCarId()

        val application = ApplicationRegister(
            carId = carId,
            name = cleaned.name,
            businessOwner = cleaned.businessOwner,
            applicationManager = cleaned.applicationManager
        )
        apply(application, cleaned)
        application.createdBy = username
        application.updatedBy = username

        return ApplicationRegisterDetail.from(repository.save(application))
    }

    @Transactional
    open fun update(id: Long, request: ApplicationRegisterRequest, username: String): ApplicationRegisterDetail {
        val application = getWithAssets(id)
        val cleaned = clean(request)

        apply(application, cleaned)
        application.updatedBy = username

        return ApplicationRegisterDetail.from(repository.update(application))
    }

    @Transactional
    open fun delete(id: Long) {
        if (!repository.existsById(id)) {
            throw NoSuchElementException("Application not found")
        }
        repository.deleteById(id)
    }

    @Transactional
    open fun replaceAssets(id: Long, assetIds: List<Long>, username: String): ApplicationRegisterDetail {
        val application = getWithAssets(id)
        val distinctIds = assetIds.distinct()
        val assets = if (distinctIds.isEmpty()) emptyList() else assetRepository.findByIdIn(distinctIds)
        val foundIds = assets.mapNotNull { it.id }.toSet()
        val missing = distinctIds.filterNot { foundIds.contains(it) }
        if (missing.isNotEmpty()) {
            throw IllegalArgumentException("Unknown asset ids: ${missing.joinToString(", ")}")
        }

        application.assets.clear()
        application.assets.addAll(assets)
        application.updatedBy = username

        return ApplicationRegisterDetail.from(repository.update(application))
    }

    private fun getWithAssets(id: Long): ApplicationRegister {
        return repository.findByIdWithAssets(id).orElseThrow {
            NoSuchElementException("Application not found")
        }
    }

    private fun clean(request: ApplicationRegisterRequest): ApplicationRegisterRequest {
        val name = request.name.trim()
        val businessOwner = request.businessOwner.trim()
        val applicationManager = request.applicationManager.trim()
        val errors = mutableListOf<String>()
        if (name.isBlank()) errors.add("name is required")
        if (businessOwner.isBlank()) errors.add("businessOwner is required")
        if (applicationManager.isBlank()) errors.add("applicationManager is required")
        if (errors.isNotEmpty()) {
            throw IllegalArgumentException(errors.joinToString("; "))
        }

        return request.copy(
            name = name,
            businessOwner = businessOwner,
            applicationManager = applicationManager,
            criticality = request.criticality.clean(),
            operationalStatus = request.operationalStatus.clean(),
            applicationTechnology = request.applicationTechnology.clean(),
            applicationArchitecture = request.applicationArchitecture.clean(),
            informationClassification = request.informationClassification.clean(),
            processingOfPersonalData = request.processingOfPersonalData.clean(),
            icsRelevant = request.icsRelevant.clean(),
            applicationExportControlRelevant = request.applicationExportControlRelevant.clean(),
            operationModel = request.operationModel.clean(),
            productionOperatingHours = request.productionOperatingHours.clean(),
            serviceOperatingHours = request.serviceOperatingHours.clean(),
            backupRecoveryUrl = request.backupRecoveryUrl.clean(),
            incidentAssignmentGroup = request.incidentAssignmentGroup.clean(),
            notes = request.notes.clean(),
            cmdbWorkspaceUrl = request.cmdbWorkspaceUrl.clean()
        )
    }

    private fun apply(application: ApplicationRegister, request: ApplicationRegisterRequest) {
        application.name = request.name
        application.criticality = request.criticality
        application.operationalStatus = request.operationalStatus
        application.businessOwner = request.businessOwner
        application.applicationManager = request.applicationManager
        application.applicationTechnology = request.applicationTechnology
        application.applicationArchitecture = request.applicationArchitecture
        application.lastQualityCheck = request.lastQualityCheck
        application.informationClassification = request.informationClassification
        application.processingOfPersonalData = request.processingOfPersonalData
        application.icsRelevant = request.icsRelevant
        application.applicationExportControlRelevant = request.applicationExportControlRelevant
        application.operationModel = request.operationModel
        application.productionOperatingHours = request.productionOperatingHours
        application.serviceOperatingHours = request.serviceOperatingHours
        application.backupRecoveryUrl = request.backupRecoveryUrl
        application.incidentAssignmentGroup = request.incidentAssignmentGroup
        application.notes = request.notes
        application.cmdbWorkspaceUrl = request.cmdbWorkspaceUrl
    }

    private fun allocateCarId(): String {
        val highest = repository.findAllCarIds()
            .mapNotNull { it.toLongOrNull() }
            .maxOrNull() ?: 0L
        var candidate = highest + 1
        while (repository.existsByCarIdIgnoreCase(candidate.toString())) {
            candidate += 1
        }
        return candidate.toString()
    }

    private fun String?.clean(): String? = this?.trim()?.ifBlank { null }
}

File: src/frontend/src/services/applicationRegisterService.ts
import { authenticatedDelete, authenticatedGet, authenticatedPost, authenticatedPut } from '../utils/auth';

export interface ApplicationAssetSummary {
  id: number;
  name: string;
  type: string;
  owner: string;
  ip?: string | null;
}

export interface ApplicationRegisterSummary {
  id: number;
  carId: string;
  name: string;
  criticality?: string | null;
  operationalStatus?: string | null;
  businessOwner: string;
  applicationManager: string;
  updatedAt?: string | null;
}

export interface ApplicationRegisterDetail extends ApplicationRegisterSummary {
  applicationTechnology?: string | null;
  applicationArchitecture?: string | null;
  lastQualityCheck?: string | null;
  informationClassification?: string | null;
  processingOfPersonalData?: string | null;
  icsRelevant?: string | null;
  applicationExportControlRelevant?: string | null;
  operationModel?: string | null;
  productionOperatingHours?: string | null;
  serviceOperatingHours?: string | null;
  backupRecoveryUrl?: string | null;
  incidentAssignmentGroup?: string | null;
  notes?: string | null;
  cmdbWorkspaceUrl?: string | null;
  createdAt?: string | null;
  createdBy?: string | null;
  updatedBy?: string | null;
  assets: ApplicationAssetSummary[];
}

export type ApplicationRegisterRequest = Omit<ApplicationRegisterDetail, 'id' | 'assets' | 'createdAt' | 'updatedAt' | 'createdBy' | 'updatedBy'>;
export type ApplicationRegisterSaveRequest = Omit<ApplicationRegisterRequest, 'carId'> & { carId?: string | null };

const optionalApplicationFields: Array<keyof ApplicationRegisterSaveRequest> = [
  'criticality',
  'operationalStatus',
  'applicationTechnology',
  'applicationArchitecture',
  'lastQualityCheck',
  'informationClassification',
  'processingOfPersonalData',
  'icsRelevant',
  'applicationExportControlRelevant',
  'operationModel',
  'productionOperatingHours',
  'serviceOperatingHours',
  'backupRecoveryUrl',
  'incidentAssignmentGroup',
  'notes',
  'cmdbWorkspaceUrl',
];

function sanitizeRequest(request: ApplicationRegisterSaveRequest): ApplicationRegisterSaveRequest {
  const sanitized = { ...request };
  optionalApplicationFields.forEach((field) => {
    if (sanitized[field] === '') {
      sanitized[field] = undefined as never;
    }
  });
  return sanitized;
}

async function parseOrThrow<T>(response: Response, fallback: string): Promise<T> {
  if (response.ok) {
    return await response.json();
  }
  const body = await response.json().catch(() => ({}));
  throw new Error(body.error || body.message || `${fallback}: ${response.status}`);
}

export async function listApplications(search = ''): Promise<ApplicationRegisterSummary[]> {
  const params = new URLSearchParams();
  if (search.trim()) params.set('search', search.trim());
  const suffix = params.toString() ? `?${params.toString()}` : '';
  return parseOrThrow(await authenticatedGet(`/api/applications${suffix}`), 'Failed to load applications');
}

export async function getApplication(id: number): Promise<ApplicationRegisterDetail> {
  return parseOrThrow(await authenticatedGet(`/api/applications/${id}`), 'Failed to load application');
}

export async function createApplication(request: ApplicationRegisterSaveRequest): Promise<ApplicationRegisterDetail> {
  return parseOrThrow(await authenticatedPost('/api/applications', sanitizeRequest(request)), 'Failed to create application');
}

export async function updateApplication(id: number, request: ApplicationRegisterSaveRequest): Promise<ApplicationRegisterDetail> {
  return parseOrThrow(await authenticatedPut(`/api/applications/${id}`, sanitizeRequest(request)), 'Failed to update application');
}

export async function deleteApplication(id: number): Promise<void> {
  const response = await authenticatedDelete(`/api/applications/${id}`);
  if (!response.ok) {
    const body = await response.json().catch(() => ({}));
    throw new Error(body.error || body.message || `Failed to delete application: ${response.status}`);
  }
}

export async function replaceApplicationAssets(id: number, assetIds: number[]): Promise<ApplicationRegisterDetail> {
  return parseOrThrow(
    await authenticatedPut(`/api/applications/${id}/assets`, { assetIds }),
    'Failed to update linked assets'
  );
}
